ledger/eval/txntracer.go

```diff
@@ -0,0 +1,121 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package eval
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+)
+
+var txnGroupDeltaTracer *TxnGroupDeltaTracer
+
+// TODO do we want something like this?
+type txnGroupDeltas struct {
+	// Ids contains all the associated IDs for these changes (group and txn IDs)
+	Ids []crypto.Digest
+	ledgercore.StateDelta
+}
+
+// TxnGroupDeltaTracer collects groups of StateDelta objects covering groups of txns
+type TxnGroupDeltaTracer struct {
+	// lookback is the number of rounds stored at any given time
+	Lookback uint64
+	// no-op methods we don't care about
+	logic.NullEvalTracer
+	// txnGroupDeltas stores the StateDelta objects for each round, indexed by all the IDs within the group
+	txnGroupDeltas map[basics.Round]map[crypto.Digest]*ledgercore.StateDelta
+	// latestRound is the most recent round seen via the BeforeBlock hdr
+	latestRound basics.Round
+}
+
+// makeTxnGroupDeltaTracer creates a TxnGroupDeltaTracer
+func makeTxnGroupDeltaTracer(lookback uint64) *TxnGroupDeltaTracer {
+	return &TxnGroupDeltaTracer{
+		Lookback:       lookback,
+		txnGroupDeltas: make(map[basics.Round]map[crypto.Digest]*ledgercore.StateDelta),
+	}
+}
+
+// TxnGroupDeltaTracerForConfig retrieves the TxnGroupDeltaTracer or creates it if it does not already exist
+func TxnGroupDeltaTracerForConfig(cfg config.Local) *TxnGroupDeltaTracer {
+	if txnGroupDeltaTracer == nil {
+		txnGroupDeltaTracer = makeTxnGroupDeltaTracer(cfg.MaxAcctLookback)
+	}
+	return txnGroupDeltaTracer
+}
+
+// BeforeBlock implements the EvalTracer interface for pre-block evaluation
+func (tracer *TxnGroupDeltaTracer) BeforeBlock(hdr *bookkeeping.BlockHeader) {
+	// Drop older rounds based on the Lookback parameter
+	delete(tracer.txnGroupDeltas, hdr.Round-basics.Round(tracer.Lookback))
+	tracer.latestRound = hdr.Round
+}
+
+// AfterTxnGroup implements the EvalTracer interface for txn group boundaries
+func (tracer *TxnGroupDeltaTracer) AfterTxnGroup(ep *logic.EvalParams, deltas *ledgercore.StateDelta, evalError error) {
+	var txnDeltaMap = make(map[crypto.Digest]*ledgercore.StateDelta)
```

Comment:
nit: initialize map size to `len(ep.TxnGroup)`? Maybe add 1 for a possible group ID as well?